File: 2023/kotlin/src/Day01.kt
private fun calibrate(values: List<String>) = values.sumOf { value ->
    val digits = value.replace(Regex("\\D"), "")
    "${digits.first()}${digits.last()}".toInt()
}

private fun part1(input: String): Int = calibrate(input.lines())

private fun part2(input: String): Int {
    val translations = mapOf(
        "one" to "o1e",
        "two" to "t2o",
        "three" to "t3e",
        "four" to "4",
        "five" to "5e",
        "six" to "6",
        "seven" to "7n",
        "eight" to "e8t",
        "nine" to "n9e",
    )

    val values = input.lines().map { value ->
        translations.entries.fold(value) { acc, (from, to) ->
            acc.replace(from, to)
        }
    }

    return calibrate(values)
}

fun main() {
    val input = readInput("Day01")
    part1(input).println()
    part2(input).println()
}

File: 2023/swift/AdventOfCode2023/Sources/Day01.swift
struct Day01: AdventDay {
  var input: String

  func calibrate(_ values: [String]) -> Int {
    values.map { value in
      let digits = value.matches(of: /\d/).map { $0.output }
      return Int("\(digits.first!)\(digits.last!)")!
    }.sum()
  }

  func part1() -> Any {
    calibrate(input.lines())
  }

  func part2() -> Any {
    let translations = [
      "one": "o1e",
      "two": "t2o",
      "three": "t3e",
      "four": "4",
      "five": "5e",
      "six": "6",
      "seven": "7n",
      "eight": "e8t",
      "nine": "n9e",
    ]

    let values = input.lines().map { value in
      translations.reduce(value) { translated, entry in
        translated.replacingOccurrences(of: entry.key, with: entry.value)
      }
    }

    return calibrate(values)
  }
}
